:lang() selectors should serialize without quotation marks if they were parsed without them

darinadler opened the pull request:
#### 7e31f273a4e1bc368bb568101db621e4c9a084e2
<pre>
:lang() selectors should serialize without quotation marks if they were parsed without them
<a href="https://bugs.webkit.org/show_bug.cgi?id=247747">https://bugs.webkit.org/show_bug.cgi?id=247747</a>
rdar://problem/102198822

Reviewed by Ryosuke Niwa.

Changed so we keep a record of whether a language was specified as an identifier or a string
when parsing the lang selector. This affects serialization, otherwise ignored.

* LayoutTests/fast/css/css-lang-selector-with-string-arguments-text-expected.txt: Updated.
* LayoutTests/fast/css/css-lang-selector-with-string-arguments-text.html: Updated test to not
expect quotation marks around lang values that are unquoted identifiers.

* LayoutTests/fast/css/css-selector-text-expected.txt: Ditto.
* LayoutTests/fast/css/css-selector-text.html: Ditto.

* LayoutTests/imported/w3c/web-platform-tests/css/cssom/CSS-expected.txt: Removed.
There is no corresponding test for this expected result, presumably left behind by accident.

* LayoutTests/imported/w3c/web-platform-tests/css/cssom/CSSStyleRule-set-selectorText-expected.txt:
Expect more PASS.
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/selectorSerialize-expected.txt: Ditto.

* Source/WebCore/css/CSSSelector.cpp:
(WebCore::appendLangArgumentList): Use serializeIdentifier and serializeString for correct quoting
and select which one based on the wasQuoted flag in PossiblyQuotedIdentifier.
(WebCore::CSSSelector::selectorText const): Take an additional separator argument,
and rearranged code a bit. Also updated for PossiblyQuotedIdentifier.
(WebCore::CSSSelector::setArgumentList): New argument type.

* Source/WebCore/css/CSSSelector.h: Added PossiblyQuotedIdentifier. Initialize data members in the
class definition so we don&apos;t have to do it in all the constructors. Tweaked comment format a bit.
Changed type of argumentList. Moved inline function bodies out of the class. Deleted the assignment
operator instead of declaring it private an just not defining it. Use a FixedVector for
argumentList instead of a unique_ptr&lt;Vector&gt;.

* Source/WebCore/css/SelectorChecker.cpp:
(WebCore::SelectorChecker::checkOne const): Removed unneeded braces.

* Source/WebCore/css/SelectorCheckerTestFunctions.h:
(WebCore::matchesLangPseudoClass): Updated for change in type.
* Source/WebCore/css/parser/CSSParserSelector.cpp:
(WebCore::CSSParserSelector::setArgumentList): Ditto.
* Source/WebCore/css/parser/CSSParserSelector.h: Ditto.

* Source/WebCore/css/parser/CSSSelectorParser.cpp:
(WebCore::consumePossiblyQuotedIdentifer): Added.
(WebCore::consumeLangArgumentList): Updated to return the new type, also cut down on repeated code
by using the consumePossiblyQuotedIdentifer function.
(WebCore::CSSSelectorParser::consumePseudo): Updated for the above.

* Source/WebCore/cssjit/SelectorCompiler.cpp:
(WebCore::SelectorCompiler::addPseudoClassType): Update for the new type.
(WebCore::SelectorCompiler::SelectorCodeGenerator::generateElementIsInLanguage): Ditto.

Canonical link: <a href="https://commits.webkit.org/256603@main">https://commits.webkit.org/256603@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8bf9eb88edcc3409b0e75362e93147b202f591c8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/96303 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/5557 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/29358 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/105844 "Built successfully") | [💥 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/166187 "An unexpected error occured. Recent messages:Pull request contains relevant changes; Failed to print configuration") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/100285 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/5688 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/34302 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/88681 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/102574 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/101974 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/4211 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/82898 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/31233 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/86053 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/87968 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/74071 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/40034 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/19481 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/37710 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/20847 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4579 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/131 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/43433 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/139 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/40115 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/css/CSSSelector.h:
```diff
@@ -346,25 +317,22 @@ namespace WebCore {
+        unsigned m_relation : 4 { DescendantSpace }; // enum RelationType.
+        mutable unsigned m_match : 4 { Unknown }; // enum Match.
+        mutable unsigned m_pseudoType : 8 { 0 }; // PseudoType.
+        bool m_isLastInSelectorList : 1 { false };
```
I think we need to use `unsigned` for these `bool` types for them to be packed tight in Windows.

darinadler replied:
Yes, I tested and that is still true even with the latest MSVC. Will fix.

rniwa replied:
I think they can't change it because of ABI compatibility.

rniwa commented on Source/WebCore/css/CSSSelector.h:
```diff
@@ -346,25 +317,22 @@ namespace WebCore {
+        bool m_isForPage : 1 { false };
+        bool m_tagIsForNamespaceRule : 1 { false };
+        bool m_caseInsensitiveAttributeValueMatching : 1 { false };
+        bool m_destructorHasBeenCalled : 1 { false };
```
Any reason why this has to be compiled all the time now?

darinadler replied:
I changed this back.

rniwa commented on Source/WebCore/css/CSSSelector.h:
```diff
@@ -573,10 +524,8 @@ inline CSSSelector::CSSSelector(const CSSSelector& o)
 {
-    ASSERT_WITH_SECURITY_IMPLICATION(!m_destructorHasBeenCalled);
-#if !ASSERT_WITH_SECURITY_IMPLICATION_DISABLED
+    RELEASE_ASSERT_WITH_SECURITY_IMPLICATION(!m_destructorHasBeenCalled);
```
Hm... we're making this a release assert.
Perhaps we should do it in a separate patch?

darinadler replied:
I figured out how to remove this change without upsetting the style checker. I am in no hurry to make the change, so I will just set it aside entirely for now.

rniwa replied:
Okay. Yeah, I usually ignore that style checker error when I'm dealing with an existing debug only assertion.

rniwa commented on Source/WebCore/css/CSSSelector.h:
```diff
@@ -23,46 +23,53 @@
+
+struct PossiblyQuotedIdentifier {
+    AtomString identifier;
+    bool wasQuoted { false };
```
Maybe we can use an enum class? e.g. `enum class QuoteState { NotQuoted, Quoted }`.

darinadler replied:
I’m not sure that would improve things. We normally do this to avoid passing a constant `true` or `false` in a context where it’s not clear what it means. But in this case there is nowhere doing that. There is only one call site passing this value and it passes the result of a boolean expression.

However if you are convinced it would make the code better I would do it.

rniwa replied:
Okay. I don't feel strongly about it either way.

darinadler committed:
:lang() selectors should serialize without quotation marks if they were parsed without them
https://bugs.webkit.org/show_bug.cgi?id=247747
rdar://problem/102198822

Reviewed by Ryosuke Niwa.

Changed so we keep a record of whether a language was specified as an identifier or a string
when parsing the lang selector. This affects serialization, otherwise ignored.

* LayoutTests/fast/css/css-lang-selector-with-string-arguments-text-expected.txt: Updated.
* LayoutTests/fast/css/css-lang-selector-with-string-arguments-text.html: Updated test to not
expect quotation marks around lang values that are unquoted identifiers.

* LayoutTests/fast/css/css-selector-text-expected.txt: Ditto.
* LayoutTests/fast/css/css-selector-text.html: Ditto.

* LayoutTests/imported/w3c/web-platform-tests/css/cssom/CSS-expected.txt: Removed.
There is no corresponding test for this expected result, presumably left behind by accident.

* LayoutTests/imported/w3c/web-platform-tests/css/cssom/CSSStyleRule-set-selectorText-expected.txt:
Expect more PASS.
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/selectorSerialize-expected.txt: Ditto.

* Source/WebCore/css/CSSSelector.cpp:
(WebCore::appendLangArgumentList): Use serializeIdentifier and serializeString for correct quoting
and select which one based on the wasQuoted flag in PossiblyQuotedIdentifier.
(WebCore::CSSSelector::selectorText const): Take an additional separator argument,
and rearranged code a bit. Also updated for PossiblyQuotedIdentifier.
(WebCore::CSSSelector::setArgumentList): New argument type.

* Source/WebCore/css/CSSSelector.h: Added PossiblyQuotedIdentifier. Initialize data members in the
class definition so we don't have to do it in all the constructors. Tweaked comment format a bit.
Changed type of argumentList. Moved inline function bodies out of the class. Deleted the assignment
operator instead of declaring it private an just not defining it. Use a FixedVector for
argumentList instead of a unique_ptr<Vector>.

* Source/WebCore/css/SelectorChecker.cpp:
(WebCore::SelectorChecker::checkOne const): Removed unneeded braces.

* Source/WebCore/css/SelectorCheckerTestFunctions.h:
(WebCore::matchesLangPseudoClass): Updated for change in type.
* Source/WebCore/css/parser/CSSParserSelector.cpp:
(WebCore::CSSParserSelector::setArgumentList): Ditto.
* Source/WebCore/css/parser/CSSParserSelector.h: Ditto.

* Source/WebCore/css/parser/CSSSelectorParser.cpp:
(WebCore::consumePossiblyQuotedIdentifer): Added.
(WebCore::consumeLangArgumentList): Updated to return the new type, also cut down on repeated code
by using the consumePossiblyQuotedIdentifer function.
(WebCore::CSSSelectorParser::consumePseudo): Updated for the above.

* Source/WebCore/cssjit/SelectorCompiler.cpp:
(WebCore::SelectorCompiler::addPseudoClassType): Update for the new type.
(WebCore::SelectorCompiler::SelectorCodeGenerator::generateElementIsInLanguage): Ditto.

Canonical link: https://commits.webkit.org/256603@main
